### PR TITLE
Fixed #24031 -- Added CASE expressions

### PR DESCRIPTION
**Pull request by michalmo**
I'm opening this PR for review of my implementation of CASE expressions.
It's working with tests passing on SQLite, postgres and MySQL with python 2.7, 3.2, 3.3 and 3.4 (except MySQL + python 3.2).

There are some issues however:

* API needs to be finalized. I've discussed this with @jarshwah and am currently using two classes ```SearchedCase``` and ```SimpleCase```, with an entry point class ```Case```, that returns one of those two depending on the arguments. You can see how Case expressions could be used with each of the classes and then we can drop some of the names once a decision has been made.
* No documentation - waiting on API decision.
* I added tests for the interaction of ```QuerySet.update``` with Case expressions and all the Field types, and came across some issues with ```Value``` expressions and ```DurationField```. I've fixed these in separate commits in this PR. How should I proceed with this? Make tickets for the issues? The fixes work, but I'd appreciate review from the authors of those pieces of code @mjtamlyn and @jarshwah.
* In some cases I've made the ```as_sql``` methods depend on the type of query compiler (detecting sql update). This works but needs to be better, maybe adding methods to ```SQLCompiler``` and overriding for ```SQLUpdateCompiler```. Input appreciated on this.

This is without the bulk_update code. I'll do a separate PR for that if it is wanted.

### CODE REVIEW

File: django/db/models/expressions.py [hunk 1]
```diff
@@ -538,6 +547,200 @@ def get_group_by_cols(self):
...
+    """
+    Base class for all case expressions.
+    """
+    NoDefault = object()
+
+    def __init__(self, cases=None, default=NoDefault, output_field=None):
+        super(BaseCaseExpression, self).__init__(output_field)
+        self.cases = self._parse_cases(cases)
+        if default is self.NoDefault:
+            default = None
```
Comment (jarshwah) [inline django/db/models/expressions.py]: If you assign the default as `Value(None)`, then you shouldn't need to conditionally add `self.default` in get/set_source_expressions. It should just convert to NULL sql string.

File: django/db/models/expressions.py [hunk 2]
```diff
@@ -538,6 +547,200 @@ def get_group_by_cols(self):
...
+    def value_sql(self, value, compiler, connection):
+        return compiler.compile(value)
+
+    def as_sql(self, compiler, connection):
+        if not self.cases:
+            if self.default is not None:
+                return compiler.compile(self.default)
+            return 'NULL', ()
+        predicate_sql, predicate_params = self.predicate_sql(compiler, connection)
+        result = ['CASE']
```
Comment (jarshwah) [inline django/db/models/expressions.py]: Constructing the entire string within the as_sql method departs from how other functions work. Is it possible to do something like:

```
class BaseCaseExpression(Func):
    function = None
    template = 'CASE %(simple)s %(conditions)s ELSE %(default)s END'
```

Then build up the dict required to fill in that template, and construct/return at the end? It may flow nicer, and allow 3rd party backends to modify the template without overriding the entire method.

File: django/db/models/expressions.py [hunk 3]
```diff
@@ -483,7 +485,14 @@ def __init__(self, value, output_field=None):
         self.value = value
 
     def as_sql(self, compiler, connection):
-        return '%s', [self.value]
+        val = self.value
+        if self._output_field_or_none is not None:
+            from django.db.models.sql.compiler import SQLUpdateCompiler
+            if isinstance(compiler, SQLUpdateCompiler):
```
Comment (jarshwah) [inline django/db/models/expressions.py]: Not liking this. The expression should not need to know what kind of compiler is being used here. Why is this necessary here, but hasn't been required before? Is this something that resolve_ref can handle without full knowledge of the type of compiler? Or do we need to re-add a similar get_db_prep method into expressions?

File: django/db/models/expressions.py [hunk 4]
```diff
@@ -538,6 +547,200 @@ def get_group_by_cols(self):
...
+        predicate_sql, predicate_params = self.predicate_sql(compiler, connection)
+        result = ['CASE']
+        result_params = []
+        if predicate_sql:
+            result.append(predicate_sql)
+        result_params.extend(predicate_params)
+        for condition, value in self.cases:
+            condition_sql, condition_params = self.condition_sql(condition, compiler, connection)
+            value_sql, value_params = self.value_sql(value, compiler, connection)
+            result.append('WHEN %s THEN %s' % (condition_sql, value_sql))
```
Comment (jarshwah) [inline django/db/models/expressions.py]: Should conditions be their own expressions that know how to render themselves? It could be an undocumented helper class of Case expressions:

```
class CaseCondition(Func):
    function = None
    template = 'WHEN %(condition)s THEN %(value)s'
    ...
```

It might reduce the size of the Case classes, but it might become too disconnected. Don't change it just because I've raised the idea, but it's something to think about.

File: django/db/models/expressions.py (hunk 3 above)
Comment (michalmo) [inline django/db/models/expressions.py]: Yeah, I don't like it either. That's why I mentioned it in my initial comment.
```SQLUpdateCompiler``` calls ```field.get_db_prep_save(val)``` on raw values, but it's not happening with Value objects.
So ```qs.update(field=value)``` works but ```qs.update(field=Value(value))``` fails for ```Field```s that require conversion. I've added test cases to ```tests.expressions``` that fail without this.
```UUIDField``` is problematic for MySQL.
```UUIDField```, ```TimeField``` and ```DurationField``` are problematic for SQLite.
I did the compiler type check here, since I'm not sure values should be processed with ```get_db_prep_save``` in the other compiler types.

**Comment by jarshwah** [comment]
These test cases are really thorough, nice work. I've made some quick notes just by looking over the code, but I'll be able to go over this more thoroughly from Sunday or Monday.

File: django/db/models/expressions.py (hunk 4 above)
Comment (michalmo) [inline django/db/models/expressions.py]: Yes, I was thinking about a Q wrapper expression.

**Comment by jarshwah** [comment]
The ticket tracking this work is https://code.djangoproject.com/ticket/24031 if you want to assign yourself in Trac. Also, if you change the name of the pull request to "Fixed #24031 -- Added Case Expressions" then this PR will be linked from the ticket.

**Comment by jarshwah** [comment]
buildbot, test this please.

**Comment by mjtamlyn** [comment]
I've not reviewed the whole work, but the changes to the `DurationField` look like a necessary fix. As this is a fairly big PR and they're a pretty simple bug, it would be beneficial to strip them into a separate ticket and PR I think.

File: django/db/models/expressions.py (hunk 3 above)
Comment (jarshwah) [inline django/db/models/expressions.py]: I haven't had time to test this myself (I will do so tonight), but can we just call `self.value = self.output_field.get_db_prep_value(self.value)` from the Value.resolve_ref method, no matter if the value is being saved or not? I'm not sure I understand the differences between get_db_save and get_db_value though. It seems weird that there are both methods anyway.

Preparing the value for use in SQL shouldn't be different between SELECT and UPDATE, but there may be cases that I'm unaware of that do require differentiation.

Comment (jarshwah) [inline django/db/models/expressions.py]: OK I've had a play around, and these are my findings so far.

Value.as_sql:

```
def as_sql(self, compiler, connection):
        val = self.value
        if self._output_field_or_none is not None:
            val = self.output_field.get_db_prep_value(val, connection=connection)
        return '%s', [val]
```

This appears to work fine for all backends. No need to differentiate between get_db_prep_{save/value} that I can find. Were there any particular examples that required you to use one but not the other?

BaseCaseExpression.as_postgresql:

- I got rid of the as_postgresql completely, and there are failing tests. The tests that fail are types that look like strings, but represent things like Binary or IPAddress.

- I got rid of the check for Update, to apply the cast to all queries. It fails on CharFields (because varchar(None) isn't a valid type. This highlighted another issue - if we were updating CharFields and the Values have an output_field, then it fails with the same error of varchar(None):

```
CaseTestModel.objects.filter(string__in=['1', '2']).update(
            string=self.create_expression(
                'integer',
                [(Value(1), Value('1', output_field=models.CharField())),
                 (Value(2), Value('2', output_field=models.CharField()))]))
```

Any datatype that requires extra information, like max_length or even max_digits for DecimalField will see similar issues.

What I think we need to do for the BaseCaseExpression.as_postgresql, is to apply casting only to fields that require it, and to do it whether it's a select or an update (do not discriminate). 

Maybe we can get get_db_prep to ask the backend to provide casting info for fields that require it - like UUID or GenericIPAddress. In the short term maybe we could document certain fields as not supported with certain expressions.

But this is scope creep, and should probably be handled or discussed separately. If anyone has any input here it would be helpful. If I'm not being clear please speak up too.

Comment (jarshwah) [inline django/db/models/expressions.py]: OK - I took a go at fixing the problems here: https://gist.github.com/jarshwah/76dbc87577b7fec05807

There were a tonne of errors on Oracle, mostly due to the fact that oracle stores the empty string as null, and returns the empty string from null. Our docs suggest not using null=True on CharFields, or really anything that derives from CharFields.

The modifications to `empty_strings_allowed` on GenericIPAddress and FilePathField are bugs, and I'm going to open tickets for each of them separately, so no need to include those lines in this PR.

Comment (michalmo) [inline django/db/models/expressions.py]: I've researched the ```Value.as_sql``` topic a bit.
The documentation definitely states that there is a difference in usage between ```get_db_prep_value``` and ```get_db_prep_save```.
https://docs.djangoproject.com/en/1.7/ref/models/fields/#django.db.models.Field.get_db_prep_save
https://docs.djangoproject.com/en/1.7/howto/custom-model-fields/#converting-query-values-to-database-values

Browsing the code, I think ```DecimalField``` highlights the difference the best. For saving, the value is converted to a ```Decimal``` with the required precision. For querying, it remains unchanged, which makes sense since when querying a ```DecimalField(decimal_places=0)``` a value of ```0.5``` should be accepted as is, but should not match anything.

So in my opinion this needs to stay, but needs to be delegated to the compiler class.
I'm also thinking that for SELECT this should be ```get_db_prep_loopup('exact', value, connection)``` instead of ```get_db_prep_value(value, connection)```.

Comment (jarshwah) [inline django/db/models/expressions.py]: A test case for DecimalField will help to narrow down what will be needed. The get_db_prep_save is going to be problematic for Value(decimal) though, since we don't require (and recently have actively recommended against) setting a max_digits/decimal_places. See https://github.com/django/django/commit/267a1dcd9b7877d97917ba6222fd247da060f9b0

If we have to differentiate between save and select, then we should try to pass that information via resolve_ref. Any test cases that will not pass due to decisions made earlier should be pulled out, and tickets created in trac with those test cases attached. These can be fixed after the alpha freeze, since they are not caused by your patch here.

**Comment by jarshwah** [comment]
@mjtamlyn @michalmo I've extracted the duration field commit into a separate PR: https://github.com/django/django/pull/3851

File: django/db/backends/__init__.py [hunk 5]
```diff
@@ -823,6 +823,15 @@ def cache_key_culling_sql(self):
         """
         return "SELECT cache_key FROM %s ORDER BY cache_key LIMIT 1 OFFSET %%s"
 
+    def case_cast_sql(self, db_type, internal_type):
```
Comment (carljm) [inline django/db/backends/__init__.py]: Is the cast SQL returned by this method actually specific to `CASE` statements in some way? Are values in `CASE` statements cast differently from other casts in SQL?

Comment (carljm) [inline django/db/backends/__init__.py]: (To make the implication explicit: if not, perhaps this method shouldn't have `case` in its name.)

File: django/db/models/aggregates.py [hunk 6]
```diff
@@ -14,9 +14,9 @@ class Aggregate(Func):
     contains_aggregate = True
     name = None
 
-    def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize=False):
+    def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize=False, for_save=False):
         assert len(self.source_expressions) == 1
-        c = super(Aggregate, self).resolve_expression(query, allow_joins, reuse, summarize)
+        c = super(Aggregate, self).resolve_expression(query, allow_joins, reuse, summarize, False)
```
Comment (carljm) [inline django/db/models/aggregates.py]: It would be helpful to have a comment here explaining why the received value of `for_save` is discarded instead of passed on to `super` like the other arguments.

Comment (jarshwah) [inline django/db/models/aggregates.py]: I noted this too and wasn't sure if it was an error or not. Can you 'save' an aggregate?

```
Model.objects.update(salary=Max('salary')) # would this work?
```

File: django/db/backends/__init__.py (hunk 5 above)
Comment (jarshwah) [inline django/db/backends/__init__.py]: Instead of passing in `db_type` and `internal_type`, I would suggest just passing the field and the connection. That way, the method can extract the pieces of information it needs, which may be different for different backends.

File: tests/expressions_case/tests.py [hunk 7]
```diff
@@ -0,0 +1,897 @@
...
+                [(Value(1), Value(datetime(2015, 1, 1))),
+                 (Value(2), Value(datetime(2015, 1, 2)))]))
+
+        self.assertQuerysetEqual(
+            CaseTestModel.objects.all().order_by('pk'),
+            [(1, datetime(2015, 1, 1)), (2, datetime(2015, 1, 2)), (3, None), (2, datetime(2015, 1, 2)), (3, None),
+             (3, None), (4, None)],
+            transform=attrgetter('integer', 'date_time'))
+
+    def test_update_decimal(self):
```
Comment (jarshwah) [inline tests/expressions_case/tests.py]: This test should include updating a decimal field that includes decimal points:

```
Value(Decimal("1.1"))
```

File: django/db/backends/__init__.py (hunk 5 above)
Comment (michalmo) [inline django/db/backends/__init__.py]: Only postgres needs this, and the existing field_cast_sql was not working for this, even though it supports GenericIPAddressField. So I added a different method.
I've skimmed through the postgres docs, and I'm not quite sure if postgres needs it due to type resolution of CASE expressions, in which case it's specific to expressions like CASE and UNION. Or if it's because this cast is one of the disallowed implicit casts for UPDATE, in which case it's required for all casts of this type in UPDATE.

I used the same arguments as ``field_cast_sql``.

File: django/db/models/aggregates.py (hunk 6 above)
Comment (michalmo) [inline django/db/models/aggregates.py]: That's a test case I missed ;) I'll check this. But now that you mention it, if this does work, I think the ``for_save`` would need to be passed to Min, and Max since they return the expression value.

**Comment by jarshwah** [comment]
It'd be helpful if there were tests covering this ticket also: https://code.djangoproject.com/ticket/11305

A measurement of success will be that case expressions can be used to implement 'conditional aggregates'.

**Comment by jarshwah** [comment]
buildbot, test this please.

**Comment by jarshwah** [comment]
Solid patch. I'm fairly happy with the implementation. I'm not too sure whether we should expose Simple and Searched cases publicly or not, so I'd appreciate input from the community on what they think.

I haven't thoroughly reviewed code style (though rough overview is fine) yet, but will do so in the next day or two. I've only casually read the docs, but the theme is mostly ok IMO.

File: django/db/backends/__init__.py (hunk 5 above)
Comment (akaariai) [inline django/db/backends/__init__.py]: It seems this method will not allow for 3rd party fields to define their own rules. This might be OK for a first cut, but we might later on want to have something better.

I wonder if we could just have a cast_sql with the following API:
```
def cast_sql(self, input_type, output_type):
    """
    Returns a cast expression from input_type to output_type. For example, if input_type is INTEGER
    and output_type is FLOAT, then returns CAST(%s AS FLOAT) for PostgreSQL.

    The return string must have %s placeholder. This will be filled with the input expression,
    which is of type input_type.
    """

Comment (akaariai) [inline django/db/backends/__init__.py]: I definitely do think we should at least try introducing the cast_sql method as described here. The argument types would be Field instances (and the names should likely be changed to input_field and output_field for consitency of Expression.output_field).

**Comment by akaariai** [comment]
I don't like the readability of simplecase. SearchedCase seems ok (though not perfect). Could we rename searchedcase to just case and remove simplecase for now?

**Comment by michalmo** [comment]
Rebased on master. Separated into 2 commits, One with the original version with fixes, and one that changes just the API to the ``SearchedCase(When('field', then=Value(1)))`` version. So you can merge the API you guys prefer, although so far its 3 votes to 0 in favor of the new API.

I can also remove SimpleCase if you really don't like it.

**Comment by jarshwah** [comment]
I agree with @akaariai that we could probably remove SimpleCase and just have the one Case/When construct for now. If there's interest in the future, we can consider adding it then. We're not losing out on too much (SearchedCase can do everything a SimpleCase can), and it'll be easier for users to grok.

I'll begin review in about an hour, and will mark as ready to merge in the morning if there are no issues.

File: django/db/models/expressions.py [hunk 8]
```diff
@@ -305,7 +308,7 @@ def prepare_database_save(self, field):
         return self
 
     def get_group_by_cols(self):
-        if not self.contains_aggregate:
+        if not self.contains_aggregate and self.allowed_in_group_by:
```
Comment (michalmo) [inline django/db/models/expressions.py]: Commit 0c7633178fa9410f102e4708cef979b873bccb76 broke tests where a case expression referenced an aggregate. The whole case SQL was added to the group by clause, I've added an ``allowed_in_group_by`` attribute on ``ExpressionNode`` which subclasses (``When`` and ``Case``) can set to ``False`` to force processing of  child expressions instead of returning ``self``. This works but I wonder if I am using ``contains_aggregate`` correctly.

File: django/db/backends/__init__.py (hunk 5 above)
Comment (jarshwah) [inline django/db/backends/__init__.py]: Michal and I discussed this on IRC and I don't think such a generic function is going to help here. This is a very narrow use case (implicit casting isn't working for certain fields in case expressions), and it needs to be special cased for certain fields. This isn't an Integer->Float kind of situation.

I'm happy enough with the method to stay. If we can generalise later then that will be OK, but I don't want this to hold up the merge.

File: django/db/backends/__init__.py [hunk 9]
```diff
@@ -823,6 +823,14 @@ def cache_key_culling_sql(self):
         """
         return "SELECT cache_key FROM %s ORDER BY cache_key LIMIT 1 OFFSET %%s"
 
+    def case_cast_sql(self, output_field):
```
Comment (michalmo) [inline django/db/backends/__init__.py]: I've tested the failing cases removing just the case expression to see if this isn't an interaction of fields with Value objects, or some other part of the code, but they don't fail. I've also looked at the types seen by the database. The fields start as ``unknown`` and the unified type is [resolved to ``text``](http://www.postgresql.org/docs/9.4/static/typeconv-union-case.html) as documented.

I've changed the method to accept the field instance, I can change the name to something like ``unification_cast_sql`` if that makes more sense.

File: django/db/models/expressions.py [hunk 10]
```diff
@@ -582,6 +593,126 @@ def get_group_by_cols(self):
...
+        self.default = default if hasattr(default, 'resolve_expression') else F(default)
+
+    def _check_cases(self, cases):
+        for case in cases:
+            if not isinstance(case, When):
+                raise TypeError("Positional argument is not a When object.")
+            if not hasattr(case.condition, 'resolve_expression'):
+                raise TypeError("Positional argument is not an expression.")
+        if not all(isinstance(case.condition, Q) for case in cases):
+            raise TypeError("SearchedCase conditions must be lookups.")
```
Comment (jarshwah) [inline django/db/models/expressions.py]: `SearchedCase` is no longer it's own type so this should be named `Case conditions must be lookups`. Alternatively, do this checking directly inside the `When` types, since users will be using these directly.

File: docs/ref/models/conditional-expressions.txt [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+                                        choices=ACCOUNT_TYPE_CHOICES,
+                                        default=REGULAR)
+
+When
+----
+
+.. class:: When(condition=None, then=Value(None), **lookups)
+
+A ``When()`` object is used to encapsulate a condition and its result for use
+in the conditional expressions. Using a ``When()`` object is similar to using
```
Comment (jarshwah) [inline docs/ref/models/conditional-expressions.txt]: `in the conditional expression.` Drop the 's' from 'expressions'

**Comment by jarshwah** [comment]
I haven't gone over the docs in great detail or the tests in great detail yet. Will try to get through these in the next few hours though. The implementation LGTM though.

File: django/db/models/expressions.py (hunk 10 above)
Comment (michalmo) [inline django/db/models/expressions.py]: Good catch, was going to remove this after I cleaned up When and forgot. ``When`` already makes sure condition is always a ``Q`` object.

File: django/db/backends/__init__.py (hunk 9 above)
Comment (jarshwah) [inline django/db/backends/__init__.py]: I think `unification_cast_sql` is a better name since it implies usefulness elsewhere. Adding a link to http://www.postgresql.org/docs/9.4/static/typeconv-union-case.html in the comments of the postgres implementation will also be useful.

File: docs/ref/models/expressions.txt [hunk 12]
```diff
@@ -328,6 +328,15 @@ instantiating the model field as any arguments relating to data validation
 (``max_length``, ``max_digits``, etc.) will not be enforced on the expression's
 output value.
 
+Conditional expressions
+-----------------------
+
+.. versionadded:: 1.8
+
+Conditional expressions allow you to use :keyword:`if` ... :keyword:`elif` ...
+:keyword:`else` logic in queries. Django supports SQL simple and searched
```
Comment (jarshwah) [inline docs/ref/models/expressions.txt]: simple case expressions are no longer supported since removal of SimpleCase right?

File: django/db/backends/__init__.py (hunk 9 above)
Comment (carljm) [inline django/db/backends/__init__.py]: Agreed re naming.

File: django/db/models/expressions.py [hunk 13]
```diff
@@ -485,12 +485,23 @@ def __init__(self, value, output_field=None):
         self.value = value
 
     def as_sql(self, compiler, connection):
-        if self.value is None:
+        val = self.value
+        if self._output_field_or_none is not None:
+            if self.for_save:
```
Comment (akaariai) [inline django/db/models/expressions.py]: If the only use of for_save is in here, we could just have it as an attribute of the compiler. (Insert/UpdateCompiler has a class-level attribute for_save=True, other compilers have it as False)

File: django/db/models/expressions.py [hunk 14]
```diff
@@ -485,12 +485,23 @@ def __init__(self, value, output_field=None):
         self.value = value
 
     def as_sql(self, compiler, connection):
-        if self.value is None:
+        val = self.value
+        if self._output_field_or_none is not None:
```
Comment (akaariai) [inline django/db/models/expressions.py]: There should be a comment why the check is against _output_field_or_none instead of output_field.

File: django/db/models/aggregates.py [hunk 15]
```diff
@@ -14,9 +14,11 @@ class Aggregate(Func):
     contains_aggregate = True
     name = None
 
-    def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize=False):
+    def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize=False, for_save=False):
+        if for_save:
```
Comment (akaariai) [inline django/db/models/aggregates.py]: This check of for_save could be added to the place where resolve_expression is called in the compiler (first call resolve_expression, then check if resolved.contains_aggregate -> FAIL).

File: django/db/models/expressions.py (hunk 13 above)
Comment (michalmo) [inline django/db/models/expressions.py]: That's what I wanted to do originally. However not every ``Value`` in an update is for saving. For example in ``When`` only the result expression needs to get ``for_save`` from the parent expression, the condition expression always gets ``for_save=False``.

**Comment by timgraham** [comment]
You can apply my edits with:
```
$ wget http://dpaste.com/0EHGW8Y.txt
$ patch -p1 -i 0EHGW8Y.txt
```
I started updating the tests to use the same bracket/indentation style as the `expressions` tests as I think it's easier to read. Could you finish the effort? If you don't have time to do so in the next 12 hours, just let me know. Overall, very nice work on the docs!

**Comment by timgraham** [comment]
Forgot to mention that adding a ``__repr__()`` for these objects would be nice, e.g.
```
>>> When(account_type=Client.GOLD, then=F('name'))
<django.db.models.expressions.When object at 0x7f906234f910>
```

File: tests/expressions_case/tests.py [hunk 16]
```diff
@@ -0,0 +1,1136 @@
...
+            Client.objects.all().order_by('pk'),
+            [('Jane Doe', 'G'), ('James Smith', 'R'), ('Jack Black', 'P')],
+            transform=attrgetter('name', 'account_type')
+        )
+
+    def test_conditional_aggregation_example(self):
+        Client.objects.create(
+            name='Jean Grey',
+            account_type=Client.REGULAR,
+            registered_on=date.today())
```
Comment (timgraham) [inline tests/expressions_case/tests.py]: move ) to next line for consistency

File: tests/expressions_case/tests.py [hunk 17]
```diff
@@ -0,0 +1,1136 @@
...
+        )
+
+
+class CaseDocumentationExamples(TestCase):
+    @classmethod
+    def setUpTestData(cls):
+        Client.objects.create(
+            name='Jane Doe',
+            account_type=Client.REGULAR,
+            registered_on=date.today() - timedelta(days=36)
```
Comment (timgraham) [inline tests/expressions_case/tests.py]: If you have time, we also try to include on trailing comma on the last kwarg so if more items are added later, we don't have to modify that line again.

File: tests/expressions_case/tests.py [hunk 18]
```diff
@@ -0,0 +1,1136 @@
...
+
+    def test_update(self):
+        CaseTestModel.objects.update(
+            string=Case(
+                When(integer=1, then=Value('one')),
+                When(integer=2, then=Value('two')),
+                default=Value('other')
+            )
+        )
+
```
Comment (timgraham) [inline tests/expressions_case/tests.py]: since the brackets on their own lines add whitespace, I think you could omit the blank line between queries/assertions in most cases to make this file a bit shorter.

**Comment by timgraham** [comment]
buildbot, add to whitelist.

**Comment by timgraham** [comment]
merged in 65246de7b1d70d25831ab394c4f4a75813f629fe, thanks!

**Comment by shaib** [comment]
buildbot, test this please.
